compiler/rustc_mir_transform/src/check_pointers.rs

```diff
@@ -0,0 +1,198 @@
+use rustc_hir::lang_items::LangItem;
+use rustc_index::IndexVec;
+use rustc_middle::mir::visit::{MutatingUseContext, NonMutatingUseContext, PlaceContext, Visitor};
+use rustc_middle::mir::*;
+use rustc_middle::ty::{self, Ty, TyCtxt};
+use tracing::{debug, trace};
+
+/// Details of a pointer check, the condition on which we decide whether to
+/// fail the assert and an [AssertKind] that defines the behavior on failure.
+pub(crate) struct PointerCheck<'tcx> {
+    pub(crate) cond: Operand<'tcx>,
+    pub(crate) assert_kind: Box<AssertKind<Operand<'tcx>>>,
+}
+
+/// Utility for adding a check for read/write on every sized, unsafe pointer.
+///
+/// Visits every read/write access to a [Sized], unsafe pointer and inserts a
+/// new basic block directly before the pointer access. Then calls `on_finding`
+/// to insert the actual logic for a pointer check (e.g. check for alignment).
+/// This utility takes care of the right order of blocks, the only thing a
+/// caller must do in `on_finding` is:
+/// - Append [Statement]s to `stmts`.
+/// - Append [LocalDecl]s to `local_decls`.
+/// - Return a [PointerCheck] that contains the condition and an [AssertKind].
+///   The AssertKind must be a panic with `#[rustc_nounwind]`.
+/// This utility will insert a terminator block that asserts on the condition
+/// and panics on failure.
+pub(crate) fn check_pointers<'a, 'tcx, F>(
```

Comment:
I assume by "unsafe pointer" you mean raw pointer? Maybe use standard terminology instead.

Reply:
Yes, makes sense. I guess this originates from `Ty::is_unsafe_ptr`, which basically boils down to a raw pointer. Does it make sense to rename that method as well?